### PR TITLE
many: expose AppStream IDs (AKA common ID)

### PR DESCRIPTION
**Pull request by bboozzoo**
The PR introduces AppStream IDs and makes them available through the local API.
 
Discussion: https://forum.snapcraft.io/t/support-for-appstream-id/2327

### CODE REVIEW

File: client/packages.go [hunk 1]
```diff
@@ -58,6 +58,7 @@ type Snap struct {
 	Broken           string        `json:"broken,omitempty"`
 	Contact          string        `json:"contact"`
 	License          string        `json:"license,omitempty"`
+	CommonIds        []string      `json:"common-ids,omitempty"`
```
Comment (zyga) [inline client/packages.go]: Should this be at a snap level or at an app level?

Comment (chipaca) [reply client/packages.go]: both :-)

File: snap/info.go [hunk 2]
```diff
@@ -636,7 +639,8 @@ type AppInfo struct {
 	// TODO: this should go away once we have more plumbing and can change
 	// things vs refactor
 	// https://github.com/snapcore/snapd/pull/794#discussion_r58688496
-	BusName string
+	BusName  string
+	CommonId string
```
Comment (zyga) [inline snap/info.go]: I guess that as long as this common-id (at app-level) is serialised and available to clients then I'm happy.

**Comment by pedronis** [comment]
@chipaca @bboozzoo is there are way to write a spread test for this?

File: snap/info.go [hunk 3]
```diff
@@ -195,6 +195,9 @@ type Info struct {
 	Tracks []string
 
 	Layout map[string]*Layout
+
+	// The list of common-ids from all apps of the snap
+	CommonIds []string
```
Comment (pedronis) [inline snap/info.go]: CommonIDs

File: snap/info.go (hunk 2 above)
Comment (pedronis) [reply snap/info.go]: CommonID also I think it should go in the grouping with Name

File: snap/info_snap_yaml.go [hunk 4]
```diff
@@ -369,6 +371,10 @@ func setAppsFromSnapYaml(y snapYaml, snap *Info) error {
 				Timer: yApp.Timer,
 			}
 		}
+		// collect all common IDs
+		if app.CommonId != "" {
+			snap.CommonIds = append(snap.CommonIds, app.CommonId)
```
Comment (pedronis) [inline snap/info_snap_yaml.go]: do we know if the same common-id can be repeated ?  should we list it only once?

File: store/details_v2.go [hunk 5]
```diff
@@ -54,6 +54,8 @@ type storeSnap struct {
 
 	// media
 	Media []storeSnapMedia `json:"media"`
+
+	CommonIds []string `json:"common-ids"`
```
Comment (pedronis) [inline store/details_v2.go]: CommonIDs

**Comment by bboozzoo** [comment]
@pedronis I think we can publish a test-snapd-* snap with proper data and then dump data directly from snapd local API

**Comment by zyga** [comment]
@bboozzoo I applied the rename so please pull in case you wanted to force push

File: snap/info.go [hunk 6]
```diff
@@ -636,7 +639,8 @@ type AppInfo struct {
 	// TODO: this should go away once we have more plumbing and can change
 	// things vs refactor
 	// https://github.com/snapcore/snapd/pull/794#discussion_r58688496
-	BusName string
+	BusName  string
+	CommonID string
```
Comment (pedronis) [inline snap/info.go]: I think it should go into the Name field grouping

File: snap/info_snap_yaml.go (hunk 4 above)
Comment (pedronis) [reply snap/info_snap_yaml.go]: result of discussion is that it should not be repeated and we need to validate for that

**Comment by pedronis** [comment]
> @pedronis I think we can publish a test-snapd-* snap with proper data and then dump data directly from snapd local API

yes, creating a test-snapd-* snap with a couple of common-ids   (probably like io.snapcraft.test-snapd-foo.bar)  seems what we need

File: snap/info.go (hunk 6 above)
Comment (bboozzoo) [reply snap/info.go]: Moved

**Comment by bboozzoo** [comment]
Published a new snap `test-snapd-appstreamid`, will need to talk to @mvo5 or store people to have it transferred to the `canonical` account.

File: snap/validate.go [hunk 7]
```diff
@@ -781,3 +786,16 @@ func ValidateLayout(layout *Layout, constraints []LayoutConstraint) error {
 	}
 	return nil
 }
+
+func ValidateCommonIDs(info *Info) error {
+	seen := make(map[string]bool, len(info.Apps))
```
Comment (pedronis) [inline snap/validate.go]: could possibly remember the name of the first appearance

File: snap/validate.go [hunk 8]
```diff
@@ -781,3 +786,16 @@ func ValidateLayout(layout *Layout, constraints []LayoutConstraint) error {
...
+	for _, app := range info.Apps {
+		if app.CommonID != "" {
+			if seen[app.CommonID] {
+				return fmt.Errorf("application %q common-id %q is not unique", app.Name, app.CommonID)
```
Comment (pedronis) [inline snap/validate.go]: "application %q common-id %q must be unique"

plus possibly

", already used by application %q"

Comment (bboozzoo) [reply snap/validate.go]: Yes, that's a nice touch.

**Comment by pedronis** [comment]
> Published a new snap test-snapd-appstreamid, will need to talk to @mvo5 or store people to have it transferred to the canonical account.

@bboozzoo for a transfer you will need store admins

**Comment by zyga** [comment]
I'm marking this as blocked as it is being re-worked for some comments not on gituhb.

**Pull request merged**